### PR TITLE
test: Code Coverage: Add Edit Expense #8

### PR DESCRIPTION
**Pull request by jayfyle**
### <samp>🤖 Generated by Copilot at 8e325c0</samp>

This pull request improves the testing and functionality of the add-edit-expense component by adding and updating mock data, test cases, and code logic. It also refactors and cleans up some code in `add-edit-expense.page.ts` for better readability and maintainability.

### <samp>🤖 Generated by Copilot at 8e325c0</samp>

> _Sing, O Muse, of the skillful coder who wrought_
> _Many changes to the add-edit-expense component,_
> _That he might test and refactor it with ease and grace,_
> _And please the gods of code review with his diligence._

### <samp>🤖 Generated by Copilot at 8e325c0</samp>

*  Extract common logic from `setupFormInit` and `setForm` functions into separate functions for getting and setting selected values for project, category, report, payment modes, and cost center in `add-edit-expense.page.ts` ([link](https://github.com/fylein/fyle-mobile-app/pull/2197/files?diff=unified&w=0#diff-492aac4c505a9ef8fb102b40f692554c3d0dc493b7abddc97b6665166efcc83bL1244-R1245), [link](https://github.com/fylein/fyle-mobile-app/pull/2197/files?diff=unified&w=0#diff-492aac4c505a9ef8fb102b40f692554c3d0dc493b7abddc97b6665166efcc83bL1270-R1273), [link](https://github.com/fylein/fyle-mobile-app/pull/2197/files?diff=unified&w=0#diff-492aac4c505a9ef8fb102b40f692554c3d0dc493b7abddc97b6665166efcc83bL1280-R1285), [link](https://github.com/fylein/fyle-mobile-app/pull/2197/files?diff=unified&w=0#diff-492aac4c505a9ef8fb102b40f692554c3d0dc493b7abddc97b6665166efcc83bL1304-R1318), [link](https://github.com/fylein/fyle-mobile-app/pull/2197/files?diff=unified&w=0#diff-492aac4c505a9ef8fb102b40f692554c3d0dc493b7abddc97b6665166efcc83bL1334-R1335), [link](https://github.com/fylein/fyle-mobile-app/pull/2197/files?diff=unified&w=0#diff-492aac4c505a9ef8fb102b40f692554c3d0dc493b7abddc97b6665166efcc83bL1348-R1362), [link](https://github.com/fylein/fyle-mobile-app/pull/2197/files?diff=unified&w=0#diff-492aac4c505a9ef8fb102b40f692554c3d0dc493b7abddc97b6665166efcc83bL1357-R1374), [link](https://github.com/fylein/fyle-mobile-app/pull/2197/files?diff=unified&w=0#diff-492aac4c505a9ef8fb102b40f692554c3d0dc493b7abddc97b6665166efcc83bL1389-R1444), [link](https://github.com/fylein/fyle-mobile-app/pull/2197/files?diff=unified&w=0#diff-492aac4c505a9ef8fb102b40f692554c3d0dc493b7abddc97b6665166efcc83bL1683-R1729))
*  Rename `setupFormInit` function to `getSelectedProjects` to reflect its purpose in `add-edit-expense.page.ts` ([link](https://github.com/fylein/fyle-mobile-app/pull/2197/files?diff=unified&w=0#diff-492aac4c505a9ef8fb102b40f692554c3d0dc493b7abddc97b6665166efcc83bL1244-R1245))
*  Remove unused parameters and observables from `setupFormInit` and `setForm` functions in `add-edit-expense.page.ts` ([link](https://github.com/fylein/fyle-mobile-app/pull/2197/files?diff=unified&w=0#diff-492aac4c505a9ef8fb102b40f692554c3d0dc493b7abddc97b6665166efcc83bL1419), [link](https://github.com/fylein/fyle-mobile-app/pull/2197/files?diff=unified&w=0#diff-492aac4c505a9ef8fb102b40f692554c3d0dc493b7abddc97b6665166efcc83bL1444))
*  Comment out hardware back button action subscription in `add-edit-expense.page.ts` as it is not relevant for testing ([link](https://github.com/fylein/fyle-mobile-app/pull/2197/files?diff=unified&w=0#diff-492aac4c505a9ef8fb102b40f692554c3d0dc493b7abddc97b6665166efcc83bL2361-R2403))
*  Add test cases for editing an expense with a corporate credit card group id in `add-edit-expense-4.spec.ts` ([link](https://github.com/fylein/fyle-mobile-app/pull/2197/files?diff=unified&w=0#diff-f43cd5c38b66131b82307192d0e690768498b3d43e4b378f36d321f42487ab57R989-R1102))
*  Add test cases for saving an edited expense with a valid or invalid form in `add-edit-expense-2.spec.ts` ([link](https://github.com/fylein/fyle-mobile-app/pull/2197/files?diff=unified&w=0#diff-c294271c8dad15915f84fbf86ca0233e9317d1c66b356eaef309058076e7fa15R468-R497))
*  Add test case for receipt parsing error scenario in `add-edit-expense-2.spec.ts` ([link](https://github.com/fylein/fyle-mobile-app/pull/2197/files?diff=unified&w=0#diff-c294271c8dad15915f84fbf86ca0233e9317d1c66b356eaef309058076e7fa15R319-R330))
*  Add mock data objects for cost centers, org user settings, report options, and unflattened expenses in `cost-centers.data.ts`, `org-user-settings.data.ts`, `report-unflattened.data.ts`, and `unflattened-txn.data.ts` ([link](https://github.com/fylein/fyle-mobile-app/pull/2197/files?diff=unified&w=0#diff-7de0329a6c26862e485c2274615a2f69250e4144bdf3abd27cc61d704384ec91R206-R221), [link](https://github.com/fylein/fyle-mobile-app/pull/2197/files?diff=unified&w=0#diff-de467ea1b214e32a89452896f193a5769a02286b73795b8723305087a9ad59c9R669-R676), [link](https://github.com/fylein/fyle-mobile-app/pull/2197/files?diff=unified&w=0#diff-d67cba8ad3a3712af6f510fa26bd7076e3e8bcd472aef15dcd35f0a4cdc32eb2R1534-R1557), [link](https://github.com/fylein/fyle-mobile-app/pull/2197/files?diff=unified&w=0#diff-21ae928ce640ad62cfada5fee7bf0bbc490300ab04a25f8309a9878e21b2f1cfR1674-R1741))
*  Add imports for `throwError`, `editExpTxn`, `By`, `unflattenedExpenseWithCCCGroupId`, `getElementBySelector`, `getElementRef`, `unflattenedData`, and `apiExpenseRes` in `add-edit-expense-2.spec.ts` and `add-edit-expense-4.spec.ts` to be used in testing ([link](https://github.com/fylein/fyle-mobile-app/pull/2197/files?diff=unified&w=0#diff-c294271c8dad15915f84fbf86ca0233e9317d1c66b356eaef309058076e7fa15L8-R8), [link](https://github.com/fylein/fyle-mobile-app/pull/2197/files?diff=unified&w=0#diff-c294271c8dad15915f84fbf86ca0233e9317d1c66b356eaef309058076e7fa15L19-R19), [link](https://github.com/fylein/fyle-mobile-app/pull/2197/files?diff=unified&w=0#diff-f43cd5c38b66131b82307192d0e690768498b3d43e4b378f36d321f42487ab57L5-R8), [link](https://github.com/fylein/fyle-mobile-app/pull/2197/files?diff=unified&w=0#diff-f43cd5c38b66131b82307192d0e690768498b3d43e4b378f36d321f42487ab57R20), [link](https://github.com/fylein/fyle-mobile-app/pull/2197/files?diff=unified&w=0#diff-f43cd5c38b66131b82307192d0e690768498b3d43e4b378f36d321f42487ab57R64-R66))
*  Add imports for `TestCases5`, `EllipsisPipe`, and `DependentFieldComponent` in `add-edit-expense.setup.spec.ts` to be used in testing ([link](https://github.com/fylein/fyle-mobile-app/pull/2197/files?diff=unified&w=0#diff-a763fec878d12e9f0fbac2b5382b030cd01da44167caaf1cff679ba11ffcdfbeR57-R59), [link](https://github.com/fylein/fyle-mobile-app/pull/2197/files?diff=unified&w=0#diff-a763fec878d12e9f0fbac2b5382b030cd01da44167caaf1cff679ba11ffcdfbeL205-R209), [link](https://github.com/fylein/fyle-mobile-app/pull/2197/files?diff=unified&w=0#diff-a763fec878d12e9f0fbac2b5382b030cd01da44167caaf1cff679ba11ffcdfbeR404))
*  Remove empty lines of code in `unflattened-txn.data.ts` ([link](https://github.com/fylein/fyle-mobile-app/pull/2197/files?diff=unified&w=0#diff-21ae928ce640ad62cfada5fee7bf0bbc490300ab04a25f8309a9878e21b2f1cfR883), [link](https://github.com/fylein/fyle-mobile-app/pull/2197/files?diff=unified&w=0#diff-21ae928ce640ad62cfada5fee7bf0bbc490300ab04a25f8309a9878e21b2f1cfR1402))
*  Add eslint directive to disable unsafe assignment rule in `add-edit-expense.setup.spec.ts` ([link](https://github.com/fylein/fyle-mobile-app/pull/2197/files?diff=unified&w=0#diff-a763fec878d12e9f0fbac2b5382b030cd01da44167caaf1cff679ba11ffcdfbeR1))

## Clickup
https://app.clickup.com/t/85zt9ph20

## Code Coverage
`78.22% Statements 1121/1433`
`65.58% Branches 1010/1540`
`77.64% Functions 330/425`
`78.71% Lines 1091/1386`

## UI Preview
Please add screenshots for UI changes

### CODE REVIEW

File: src/app/fyle/add-edit-expense/add-edit-expense-5.spec.ts [hunk 1]
```diff
@@ -360,5 +390,427 @@ export function TestCases5(getTestBed) {
...
+        expect(component.getReceiptCount).toHaveBeenCalledTimes(1);
+        expect(loaderService.showLoader).toHaveBeenCalledTimes(1);
+        expect(loaderService.hideLoader).toHaveBeenCalledTimes(1);
+        expect(customFieldsService.standardizeCustomFields).toHaveBeenCalledTimes(1);
+        expect(customInputsService.filterByCategory).toHaveBeenCalledOnceWith(expenseFieldResponse, 16577);
+        expect(component.setCategoryOnValueChange).toHaveBeenCalledTimes(1);
+        expect(component.parseFile).toHaveBeenCalledOnceWith('image');
+        expect(component.attachedReceiptsCount).toEqual(1);
+      }));
+    });
```
Comment (yksolanki9) [inline src/app/fyle/add-edit-expense/add-edit-expense-5.spec.ts:814]: Add couple more test cases with different combinations of field

File: src/app/fyle/add-edit-expense/add-edit-expense.page.ts [hunk 2]
```diff
@@ -1331,7 +1330,9 @@ export class AddEditExpensePage implements OnInit {
         return paymentModes[0].value;
       })
     );
+  }
 
+  getRecentProjects() {
```
Comment (yksolanki9) [inline src/app/fyle/add-edit-expense/add-edit-expense.page.ts:1335]: Return the value form here and set it inside the parent method

File: src/app/fyle/add-edit-expense/add-edit-expense.page.ts [hunk 3]
```diff
@@ -1345,7 +1346,20 @@ export class AddEditExpensePage implements OnInit {
         });
       })
     );
+  }
+
+  getRecentCostCenters() {
```
Comment (yksolanki9) [inline src/app/fyle/add-edit-expense/add-edit-expense.page.ts:1351]: Same as above

File: src/app/fyle/add-edit-expense/add-edit-expense.page.ts [hunk 4]
```diff
@@ -1345,7 +1346,20 @@ export class AddEditExpensePage implements OnInit {
...
+      costCenters: this.costCenters$,
+      recentValue: this.recentlyUsedValues$,
+    }).pipe(
+      concatMap(({ costCenters, recentValue }) =>
+        this.recentlyUsedItemsService.getRecentCostCenters(costCenters, recentValue)
+      )
+    );
+  }
 
+  getRecentCurrencies() {
```
Comment (yksolanki9) [inline src/app/fyle/add-edit-expense/add-edit-expense.page.ts:1362]: same

File: src/app/fyle/add-edit-expense/add-edit-expense.page.ts [hunk 5]
```diff
@@ -1386,13 +1402,45 @@ export class AddEditExpensePage implements OnInit {
...
+
+    const selectedPaymentMode$ = this.getSelectedPaymentModes();
+
+    this.getRecentCostCenters();
+
+    const defaultPaymentMode$ = this.getDefaultPaymentModes();
+
+    this.getRecentProjects();
+
+    this.getRecentCurrencies();
```
Comment (yksolanki9) [inline src/app/fyle/add-edit-expense/add-edit-expense.page.ts:1437]: Set the values here just like the other methods

File: src/app/fyle/add-edit-expense/add-edit-expense.setup.spec.ts [hunk 6]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable @typescript-eslint/no-unsafe-assignment */
```
Comment (yksolanki9) [inline src/app/fyle/add-edit-expense/add-edit-expense.setup.spec.ts:1]: Why is this needed?

**Review by yksolanki9 — CHANGES_REQUESTED** [review]
Good job with the refactoring 👍 Things will look much cleaner now. 
One important thing - Test it thoroughly with different combinations of expense fields. Anything breaking here is a p0

**Review by Dimple16 — COMMENTED** [review]
one review is done, removing myself

File: src/app/fyle/add-edit-expense/add-edit-expense-5.spec.ts (hunk 1 above)
Comment (jayfyle) [reply src/app/fyle/add-edit-expense/add-edit-expense-5.spec.ts:814]: Added in the upcoming PRs for 90%

**Review by yksolanki9 — APPROVED** [review]

**Pull request merged**